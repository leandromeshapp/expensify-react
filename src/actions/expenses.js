import uuid from "uuid"
import database from "../firebase/firebase"



// ADD EXPENSE
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense,
})


export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        const {
            description = "", 
            note = "", 
            amount = 0, 
            createdAt = 0 
        } = expenseData
        const expense = { description, note, amount, createdAt }
        return database.ref(`users/${uid}/expenses`).push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense,
            }))
        })
    }
}


// REMOVE EXPENSE
    export const removeExpense = id => ({
    type: "REMOVE_EXPENSE",
    id,
})


export const startRemoveExpense = id => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses/${id}`)
        .remove().then(() => {
            dispatch(removeExpense(id))
        })
    }
}




// EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
})


export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses/${id}`)
        .update(updates)
        .then(() => {
            dispatch(editExpense(id, updates));
        });
    };
  };


// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})


export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses`).once("value")
        .then((snapshot) => {
            const expenses = []

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}



// SET_order_expenses
export const setOrderExpenses = (expenses) => ({
    type: "SET_ORDER_EXPENSES",
    expenses
})


export const orderExpenses = (expenses) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses`)
        .orderByKey()
        .on("child_added", callback)
        .limitToFirst(5)
        .then(() => {
            dispatch(setOrderExpenses(expenses))
        })
    }
}