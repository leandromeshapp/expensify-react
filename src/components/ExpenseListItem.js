import React from "react"
import { Link } from "react-router-dom"
import moment from "moment"
import numeral from "numeral"
//import locale from 'numeral/locales';

//numeral.locale('pt-pt');
const ExpenseListItem = ({ id, description, amount, createdAt, note }) => (
    <Link className="list-item" to={`/edit/${id}`}>
        <div>
            <h3 className="list-item__title"> 
                { description } 
            </h3>

            <span className="list-item__sub-title"> { moment(createdAt).format("D MMMM YYYY") } </span>
        </div>
        <span className="list-item__note">{note}</span>
        <h3 className="list-item__data"> { numeral(amount / 100).format("0,0.00$") }  </h3>
    </Link>
)

export default ExpenseListItem