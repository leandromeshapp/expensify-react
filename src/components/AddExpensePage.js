import React from "react"
import ExpenseForm from "./ExpenseForm"
import { connect } from "react-redux"
import { startAddExpense } from "../actions/expenses"


export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        //props.dispatch(addExpense(expense))
        this.props.startAddExpense(expense)
        this.props.history.push("/")
    }

    render() {
        const {locale, dictionary} = this.props
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title"> {dictionary.pageAddExpense} </h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}


const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),

    locale: state.lang.locale,
    dictionary: state.lang.dictionary
})

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})


export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage)