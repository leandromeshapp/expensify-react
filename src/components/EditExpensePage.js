import React from "react"
import { connect } from "react-redux" 
import ExpenseForm from "./ExpenseForm"
import { startEditExpense, startRemoveExpense } from "../actions/expenses"
import { ModalDialog } from './ModalDialog';
import Modal from 'react-modal';
//import Prompt from './Prompt';

export class EditExpensePage extends React.Component {
    state = {
        modalOpen: false
      };
    
    openModal = () => {
        console.log('open modal');
        Modal.setAppElement(document.getElementById('app'));
        this.setState(() => ({ modalOpen: true }));
    };
    
    closeModal = () => {
        this.setState(() => ({ modalOpen: false }));
    };
    
    
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push("/")
    }

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id })
        this.props.history.push("/")
    }

    render() {
        const {locale, dictionary} = this.props
        return (  
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title"> {dictionary.pageEditExpense} </h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        expense = {this.props.expense}
                        onSubmit = {this.onSubmit}
                    />

            </div>
                <button 
                    className="button button--secondary" 
                    onClick={this.openModal}> 
                    Remove Expense 
                </button>
                

                <ModalDialog
                    modalOpen={this.state.modalOpen}
                    handleModalCancel={this.closeModal}
                    contentLabel="Confirm Remove Expense"
                >

                <h3>Do You Really Want To Remove This Expense?</h3>
                <p>This cannot be undone!</p>
                <button className="button" onClick={this.closeModal}>
                    Cancel
                </button>
                <button
                    className="button button--destructive"
                    onClick={this.onRemove}>
                    Remove
                </button>
                </ModalDialog>
        </div>
        )
    }
}


const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),

    locale: state.lang.locale,
    dictionary: state.lang.dictionary
})


const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)