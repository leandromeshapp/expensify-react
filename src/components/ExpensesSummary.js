import React from "react"
import { connect } from "react-redux"

import numeral from "numeral"
import locale from 'numeral/locales';

import selectExpenses from "../selectors/expenses"
import selectExpensesTotal from "../selectors/expenses-total"

import { Link } from "react-router-dom"
import { showAll } from "../actions/filters"

import replaceAll from '../utils/replaceAll';

export const ExpensesSummary = ({ expenses, expensesTotal, filters, visibleExpenseCount, visibleExpensesTotal, dictionary }) => {
    const expenseCount = expenses.length
    const filteredExpensesCount = expenseCount - visibleExpenseCount
    const visibleExpenseWord = visibleExpenseCount === 1 ? 'expense' : 'expenses'
    //const formattedVisibleExpensesTotal = numeral(visibleExpensesTotal / 100).format('$0,0.00')
    numeral.locale('pt-pt');
    const formattedVisibleExpensesTotal = numeral(visibleExpensesTotal / 100).format('0,0.00 $')
    // const formattedFilteredExpensesTotal = numeral((expensesTotal - visibleExpensesTotal) / 100).format('$0,0.00')
    const formattedFilteredExpensesTotal = numeral((expensesTotal - visibleExpensesTotal) / 100).format('0,0.00 $')
    
    const existingExpensesButFiltered = () => {
        if (expenseCount >= 1 && visibleExpenseCount < 1) {
          return true
        }
        return false
    };

    return (
        <div className="page-header">
            <div className="content-container">
            { expenseCount < 1 && 
                <h1 className="page-header__title"> {dictionary.summaryMessageNoExpenses} </h1>
            }

            { existingExpensesButFiltered() && 
                <h1 className="page-header__title"> {dictionary.summaryMessageNoMatch} </h1>
            }

            { visibleExpenseCount >= 1 && 
                <h1 className="page-header__title"
                    dangerouslySetInnerHTML={{
                        __html: replaceAll(dictionary.summaryMessageTitle, {
                            "{p1}": `<span>${expenseCount}</span>`,
                            "{p2}": expenseCount != 1 ? 's' : '',
                            "{p3}": `<span>${numeral(expensesTotal / 100).format('$0,0.00')}</span>`
                        })
                    }}
                >
                </h1>
            }
                
            <div className="page-header__action">
                <Link className="button button" to="/create"> {dictionary.pageAddExpense}</Link>
            </div>
            <br/>

            { expenseCount !== visibleExpenseCount &&
                <div className="header-filter-message"
                    dangerouslySetInnerHTML={{
                        __html: replaceAll(dictionary.summaryMessageSubitle, {
                            "{p1}": `<span>${filteredExpensesCount}</span>`,
             

                            "{p2}": filteredExpensesCount === 1 ? ' expense is not ' : ' expenses are not',
                                    

                            "{p3}": `<span>${formattedFilteredExpensesTotal}</span>`
                        })
                    }} 
                >
                </div>}
            </div>
        </div>
    )
} 


const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenses: state.expenses,
        expensesTotal: selectExpensesTotal(state.expenses),
        filters: state.filters,
        visibleExpenseCount: visibleExpenses.length,
        visibleExpensesTotal: selectExpensesTotal(visibleExpenses),

        locale: state.lang.locale,
        dictionary: state.lang.dictionary
    }
}

export default connect(mapStateToProps, {showAll})(ExpensesSummary)