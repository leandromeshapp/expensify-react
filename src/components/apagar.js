import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import { showAll } from "../actions/filters";
import selectExpensesTotal from "../selectors/expenses-total";
import numeral from "numeral";


export const ExpenseSummary = props => {
  const { expenses } = props;
  const expenseWord = expenses.length > 1 ? "expenses" : "expense";
  const formattedExpensesTotal = numeral(
    selectExpensesTotal(expenses) / 100
  ).format("$0,0.00");
  const filteredExpenseLength = props.filtered.length - expenses.length;
  const filteredWord = filteredExpenseLength > 1 ? "expenses" : "expense";

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenses.length} </span> {expenseWord} totalling
          <span> {formattedExpensesTotal} </span>
        </h1>

        {filteredExpenseLength !== 0 ? (
          <div className="hidden-title">
            <div className="hiddenMessage">
              Not showing{" "}
              <span className="length"> {filteredExpenseLength} </span>
              hidden {filteredWord} because of filters
            </div>
            <div>
              <button
                onClick={() => {
                  props.showAll();
                }}
                className="button button--showall"
              >
                View All
              </button>
            </div>
          </div>
        ) : (
          ""
        )}
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};
const mapStateToProps = ({ expenses, filters }) => {

  return {
    expenses: selectExpenses(expenses, filters),
    filtered: expenses
  };
};
export default connect(mapStateToProps, { showAll })(ExpenseSummary);



// WEBPACK FOOTER //
// ./src/components/ExpenseSummary.js