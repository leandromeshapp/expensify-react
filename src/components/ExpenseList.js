import React from "react"
import { connect } from "react-redux"
import { Pagination } from 'react-bootstrap';

//Components
import ExpenseListItem from "./ExpenseListItem"
import selectExpenses from "../selectors/expenses"

import { orderExpenses  } from '../actions/expenses';


import ReactPaginate from "react-paginate"


export const ExpenseList = ({ expenses, dictionary }) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile"> {dictionary.tableExpenses}</div>
            <div className="show-for-desktop"> {dictionary.tableExpense}</div>
            <div className="show-for-desktop"> {dictionary.tableAmount}</div>
        </div>
        <div className="list-body">

        {/* //<button className="buttonRegister" onClick={ orderExpenses(expenses) }> Register </button> */}
            {
                expenses.length === 0 ? (
                    <div className="list-item--message">
                        <span> No expenses </span> 
                    </div>
                    
                ) : (
                    expenses.map((expense, index) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
            }

          {/* <ReactPaginate
        previousLabel={<i className="glyphicon glyphicon-chevron-left" aria-hidden="true"></i>}
        nextLabel={<i className="glyphicon glyphicon-chevron-right" aria-hidden="true"></i>}
        breakLabel={'...'}
        pageCount={expenses.last_page}
        marginPagesDisplayed={2}
        pageRangeDisplayed={5}
        initialPage={0}
        forcePage={2 - 1}
        disableInitialCallback={true}
        //onPageChange={this.paginationPressed}
        containerClassName={"pagination"}
        subContainerClassName={"pages pagination"}
        activeClassName={"active"}
      /> */}
        </div>
    </div>
)


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),

        orderExpenses: (expenses) => dispatch(orderExpenses( expenses )),

        dictionary: state.lang.dictionary
    }
}

export default connect(mapStateToProps)(ExpenseList)