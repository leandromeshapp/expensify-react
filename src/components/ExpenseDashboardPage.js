import React from "react"
import ExpenseList from "./ExpenseList"
import ExpenseListFilters from "./ExpenseListFilters"
import ExpensesSummary from "./ExpensesSummary"
import Footer from "./Footer"

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
        <Footer />
    </div>
)

export default ExpenseDashboardPage