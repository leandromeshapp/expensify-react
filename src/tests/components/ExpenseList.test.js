import React from "react"
import { shallow } from "enzyme"
import { ExpenseList } from "../../components/ExpenseList"
import expenses from "../fixtures/expenses"



// Render expenses from ExpenseList Component
test("Should render ExpenseList with expenses correctly", () => {
    const wrapper = shallow(<ExpenseList expenses = { expenses }/>)

    expect(wrapper).toMatchSnapshot()
})



// Render ExpenseList with empty message
test("Should render ExpenseList with empty message", () => {
    const wrapper = shallow(<ExpenseList expenses = { [] }/>)

    expect(wrapper).toMatchSnapshot()
})

