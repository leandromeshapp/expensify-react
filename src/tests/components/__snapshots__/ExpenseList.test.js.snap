// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
       Expenses 
    </div>
    <div
      className="show-for-desktop"
    >
       Expense 
    </div>
    <div
      className="show-for-desktop"
    >
       Amount 
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item--message"
    >
      <span>
         No expenses 
      </span>
    </div>
  </div>
</div>
`;

exports[`Should render ExpenseList with expenses correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
       Expenses 
    </div>
    <div
      className="show-for-desktop"
    >
       Expense 
    </div>
    <div
      className="show-for-desktop"
    >
       Amount 
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={100}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note="With sugar"
    />
    <ExpenseListItem
      amount={200}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note="Without sugar"
    />
    <ExpenseListItem
      amount={300}
      createdAt={345600000}
      description="Credit Card"
      id="3"
      key="3"
      note="With chocolat"
    />
  </div>
</div>
`;
